docs/rfcs/014-storage-lsm.md

```diff
@@ -0,0 +1,146 @@
+files "archived" together into larger checkpoint files before they're
+uploaded to S3, but garbage collecting data from the archive files
+would be difficult and we have not implemented it. This proposal
+addresses that problem.
+
+
+# Overview
+
+
+```
+^ LSN
+|
+|      Memtable:     +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|
+|            L0:     +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|                    +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|                    +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|                    +-----------------------------+
+|                    |                             |
+|                    +-----------------------------+
+|
+|
+|           L1:      +-------+ +-----+ +--+  +-+
+|                    |       | |     | |  |  | |
+|                    |       | |     | |  |  | |
+|                    +-------+ +-----+ +--+  +-+
+|
+|                       +----+ +-----+ +--+  +----+
+|                       |    | |     | |  |  |    |
+|                       |    | |     | |  |  |    |
+|                       +----+ +-----+ +--+  +----+
+|
++--------------------------------------------------------------> Page ID
+
+
++---+
+|   |   Layer file
++---+
+```
+
+
+# Memtable
+
+When new WAL arrives, it is first put into the Memtable. Despite the
+name, the Memtable is not a purely in-memory data structure. It can
+spill to a temporary file on disk if the system is low on memory, and
+is accessed through a buffer cache.
+
+If the page server crashes, the Memtable is lost. It is rebuilt by
+processing again the WAL that's newer than the latest layer in L0.
+
+The size of the Memtable is equal to the "checkpoint distance", or the
+amount of WAL that we need to keep in the safekeeper.
+
+# L0
+
+When the Memtable fills up, it is written out to a new file in L0. The
+files are immutable; when a file is created, it is never
+modified. Each file in L0 is roughly 1 GB in size (*). Like the
+Memtable, each file in L0 covers the whole key range.
+
+When enough files have been accumulated in L0, compaction
+starts. Compaction processes all the files in L0 and reshuffles the
+data to create a new set of files in L1.
+
+
+(*) except in corner cases like if we want to shut down the page
+server and want to flush out the memtable to disk even though it's not
+full yet.
+
+
+# L1
+
+L1 consists of ~ 1 GB files like L0. But each file covers only part of
+the overall key space, and a larger range of LSNs. This speeds up
+searches. When you're looking for a given page, you need to check all
```

Comment:
If you don't keep a copy of all the files locally, if some of the files are only store in S3, the situation changes quite dramatically. Now reading a single page will require downloading all L0 files, so it obviously helps a lot if there are fewer of them. So that impacts recovery time, if a pageserver needs to be rebuilt.

All that said, I don't have a good idea what the constant factors here will be, i.e. at what point it's worthwhile to compact, to speed up searches. It's quite possible that the breakeven point is very high, like >> 1000 files. If that's true, then compaction should be primarily driven by garbage collection.

Reply:
> Search within one file may be O(log(n)), but if you have fewer files to search, that's faster than having to search multiple files. In other words, the search is O(m * log(n)), where 'm' is the number of files and 'n' is the number of records in each file.

Yes; but as I think I made clear in the offsite; compacting N L0-segments into N equal-sized segments in L1 still results in effectively a single L1 segment containing the LSN range of those N L0 segments, now split over N files. Sure; we can now trivially download the data at a more granular relation boundaries level at the cost of LSN granularity, but at the cost of being unable to trivially GC and compact those segments at a relation boundary and thus still having to search all data (albeit admittedly now at `n * log(m * segsize)` instead of `n * m * log(segsize)`, for n compactions stored locally, and m layers triggering a compaction).